File: apps/api/app/schemas/discovery.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from ..models import DiscoveryJobStatus, DiscoveryResultStatus, ReconciliationStatus


class DiscoveryJobCreate(BaseModel):
    provider: str = Field(default="simulated", min_length=1, max_length=50)
    target: str = Field(min_length=1, max_length=255)


class DiscoveryPageMeta(BaseModel):
    page: int
    page_size: int
    total: int


class DiscoveryJobRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    created_by_user_id: UUID | None
    provider: str
    target_type: str
    target_definition: str
    target_count: int
    status: DiscoveryJobStatus
    started_at: datetime | None
    completed_at: datetime | None
    cancelled_at: datetime | None
    error_code: str | None
    created_at: datetime
    updated_at: datetime


class DiscoveryJobPage(BaseModel):
    items: list[DiscoveryJobRead]
    meta: DiscoveryPageMeta


class DiscoveryResultRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    discovery_job_id: UUID
    organization_id: UUID
    target_ip: str
    discovered_hostname: str | None
    discovered_device_type: str | None
    discovered_operating_system: str | None
    provider: str
    status: DiscoveryResultStatus
    reconciliation_status: ReconciliationStatus
    matched_device_id: UUID | None
    discovered_at: datetime
    result_metadata: dict | None


class DiscoveryResultPage(BaseModel):
    items: list[DiscoveryResultRead]
    meta: DiscoveryPageMeta

File: apps/web/src/types/discovery.ts
/** apps/api/app/schemas/discovery.py::DiscoveryJobRead */
export type DiscoveryJob = {
  id: string
  organization_id: string
  created_by_user_id: string | null
  provider: string
  target_type: string
  target_definition: string
  target_count: number
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  started_at: string | null
  completed_at: string | null
  cancelled_at: string | null
  error_code: string | null
  created_at: string
  updated_at: string
}

/** apps/api/app/schemas/discovery.py::DiscoveryResultRead */
export type DiscoveryResult = {
  id: string
  discovery_job_id: string
  organization_id: string
  target_ip: string
  discovered_hostname: string | null
  discovered_device_type: string | null
  discovered_operating_system: string | null
  provider: string
  status: 'discovered' | 'not_reachable' | 'failed'
  reconciliation_status: 'unmatched' | 'matched' | 'conflict'
  matched_device_id: string | null
  discovered_at: string
  result_metadata: Record<string, unknown> | null
}
